backend/app/matrix/background_validator.py

```diff
@@ -0,0 +1,242 @@
+        print("Background validator has stopped")
+
+    async def start_task(self):
+        while True:
+            await self.fetch_rooms_and_users()
+
+            for room_id in self.rooms_with_conditions:
+                # When the bot is either not part of the room or dosen't have the required permissions.
+                if room_id not in self.rooms_with_permissions or room_id not in self.client.rooms:
+                    continue
+
+                ignore_members = await self.get_users_not_removed_by_bot(room_id)
+                room_specific_data = await self.client.get_account_data("rooms", room_id=room_id)
+
+                for user_id in self.registered_users.keys():
+                    if user_id in ignore_members:
+                        continue
+
+                    current_state = await self.get_current_room_membership(user_id, room_id)
+                    if current_state == "ignore":
+                        continue
+                    print(user_id, current_state)
+
+                    is_permitted_through_github = await self.check_github_conditions(
+                        user_id, room_specific_data
+                    )
+                    is_permitted = is_permitted_through_github
+
+                    next_state = await self.decide_next_state_of_user(
+                        current_state, is_permitted, room_specific_data.content.disable_room_kick
+                    )
+
+                    if current_state == next_state:
+                        continue
+
+                    await self.send_next_state_event(room_id, user_id, next_state)
+
+                await asyncio.sleep(1)
+            await asyncio.sleep(1)
+
+    async def fetch_rooms_and_users(self):
+        resp = await self.client.get_account_data("global_data")
+        self.rooms_with_conditions = resp.content.rooms
+
+        for user_id in resp.content.users:
+            user_data = await self.client.get_account_data("user", user_id=user_id)
+            self.registered_users[user_id] = RegisteredUser(user_data)
+
+        self.rooms_with_permissions = await self.client.get_rooms_with_mod_permissions(
+            self.client.user_id
+        )
+
+    async def get_current_room_membership(self, user_id: str, room_id: str) -> str:
+        matrix_room_object = self.client.rooms[room_id]
+
+        if user_id in matrix_room_object.users:
+            if matrix_room_object.users[user_id].invited:
+                return "invite"
+            else:
+                return "join"
+        else:
+            resp = await self.client.room_get_state_event(
+                room_id=room_id, event_type="m.room.member", state_key=user_id
+            )
+
+            if isinstance(resp, RoomGetStateEventError):
+                return "ignore"
+
+            membership = resp.content["membership"]
+
+            if membership == "ban":
+                return "ignore"
+            return membership
+
+    async def check_github_conditions(
```

Comment:
When a condition is added is it this method that decides what action is taken? i.e. if we ran this background task once every 30 minutes would it take 30 mins for the new condition to propagate?

Reply:
Yeah, exactly.

Reply:
We can instead set it to 10 minutes instead? Might be reasonable enough? (**The following changes will take 10 mins to take effect**)

Reply:
How hard would it be to add a way to trigger the job to run when we know it needs to? So we can have a slower running background checker but explicitly trigger it when the conditions change? i.e. we could just trigger it for a single room when that updates?

Reply:
https://github.com/MatrixDoctors/matrix-cerberus/blob/4b795406b8d321e1ad4b46c247c362e401f07a5b/backend/app/matrix/background_validator.py#L62-L97

We can try encapsulating this entire code in a method say 
```Python
def validate_room(room_id: str):
```
and call it as background_task in a path function by creating a new `BackgroundValidator` instance every time a `room_condition` is modified.

Reply:
But is there a chance that the new `BackgroundValidator` instance conflict with the long-running task?

Reply:
I think we would need to make sure only one loop was running at once, which leads me to think all the options are hard. 

You basically want a thing which is "sleep until I tell you to stop".

One possible solution to this is you have a sleep loop in between runs over all the room conditions where you check a queue of jobs ever 0.5s or similar and if there is a job in the queue run it, but otherwise keep looping until some total amount of time has expired.

Reply:
That's a great idea! I'll see what I can do.

Reply:
https://docs.python.org/3/library/asyncio-queue.html will probably be helpful.

Reply:
Thanks!

Reply:
I've added it in commit [fe7b220](https://github.com/MatrixDoctors/matrix-cerberus/pull/56/commits/fe7b2206ba932114798be7ddd41ae2a45c98276c) under PR #56